Program.cs

```diff
@@ -470,66 +472,123 @@ static P2pEndPoint GetExternalEndPoint(Socket socket)
                     Internal = (socket.LocalEndPoint as IPEndPoint)
                 };
             }
-
+            
             Console.WriteLine("Could not find a working STUN server");
+
             return null;
         }
 
-
         static int SleepTime(DateTime now)
         {
-            List<int> seconds = new List<int>() {10, 20, 30, 40, 50, 60};
-
             int next = seconds.Find(x => x > now.Second);
 
             return next - now.Second;
         }
 
-        static UdtSocket PeerConnect(Socket socket, string remoteAddr, int remotePort)
+        static async Task<UdtSocket> PeerConnectAsync(string remoteAddress, int[] remotePorts)
         {
-            bool bConnected = false;
-            int retry = 0;
+            if (!sockets.Any())
+                return null;
+
+            Console.WriteLine("Trying to connect to peer on different ports");
 
-            UdtSocket client = null;
+            UdtSocket udtSocket;
+            CancellationTokenSource cts;
+            var retry = 1;

```

Comment:
why make retry start at 1 instead of 0?

Reply:
Copy/Paste bug.

Reply:
After the `WriteLine` retry will be increased, there no problem with the code, however, it could be written in this way
```csharp
var retry = 1;
Console.WriteLine($"Try {++retry}:");
```

Reply:
you look to be talking about a refactoring?, which should be left for another PR; this PR is only to implement the 10 thread/task behaviour change

Reply:
I admit I failed the test, feel free to close the PR.

Reply:
If you keep being this negative it is when you will fail the test :) You were doing good so far ;) This is just a review, which is part of the test, hehe.

Reply:
I'm not upset with the failure. Most of the time failure is part of success. I've learned a lot from this review, thanks.